elegy/losses/cosine_similarity.py

```diff
@@ -0,0 +1,171 @@
+from elegy import types
+import typing as tp
+
+import jax.numpy as jnp
+from jax.lax import rsqrt
+
+from elegy import utils
+from elegy.losses.loss import Loss, Reduction
+
+
+def cosine_similarity(
+    y_true: jnp.ndarray, y_pred: jnp.ndarray, axis: int
+) -> jnp.ndarray:
+    """
+    Computes the cosine similarity between labels and predictions.
+
+    ```python
+    loss = -sum(l2_norm(y_true) * l2_norm(y_pred))
+    ```
+
+    Usage:
+
+    ```python
+    rng = jax.random.PRNGKey(42)
+
+    y_true = jax.random.randint(rng, shape=(2, 3), minval=0, maxval=2)
+    y_pred = jax.random.uniform(rng, shape=(2, 3))
+
+    def _l2_normalize(x, axis=None, epsilon=utils.EPSILON):
+        square_sum = jnp.sum(jnp.square(x), axis=axis, keepdims=True)
+        x_inv_norm = rsqrt(jnp.maximum(square_sum, epsilon))
+        return jnp.multiply(x, x_inv_norm)
+
+    loss = elegy.losses.cosine_similarity(y_true, y_pred, axis=1)
+
+    assert loss.shape == (2,)
+
+    y_true = _l2_normalize(y_true, axis=1)
+    y_pred = _l2_normalize(y_pred, axis=1)
+    assert jnp.array_equal(loss, -jnp.sum(y_true * y_pred, axis=1))
+    ```
+
+    Arguments:
+        y_true: Ground truth values. shape = `[batch_size, d0, .. dN]`.
+        y_pred: The predicted values. shape = `[batch_size, d0, .. dN]`.
+        axis: The dimension along which the cosinemsimilarity is computed.
+
+    Returns:
+          cosine similarity Values. If reduction is NONE, this has
+         shape [batch_size, d0, .. dN-1]; otherwise, it is scalar.
+         (Note dN-1 because all loss functions reduce by 1 dimension, usually axis=-1.)
+    """
+
+    def _l2_normalize(x, axis=None, epsilon=utils.EPSILON):
+        """Normalizes along dimension `axis` using an L2 norm."""
+        square_sum = jnp.sum(jnp.square(x), axis=axis, keepdims=True)
+        x_inv_norm = rsqrt(jnp.maximum(square_sum, epsilon))
+        return jnp.multiply(x, x_inv_norm)
+
+    y_true = _l2_normalize(y_true, axis=axis)
```

Comment:
This can simply be written as:

```python
y_true = y_true / jnp.linalg.norm(y_true, axis=axis, keepdims=True))
y_pred = y_pred / jnp.linalg.norm(y_pred, axis=axis, keepdims=True))

Reply:
Hy, @cgarciae by using this approach I'm getting sometime ```nan``` values.
for example 
```
y_true = jnp.array([[0.0, 1.0], [0.0, 0.0]]) 
y_pred = jnp.array([[0.6, 0.4], [0.4, 0.6]])
```
According to this approach if we apply CosineSimilarity in the above array we're gonna get ```nan``` values.
Till now I'm not getting what is causing the error. Kindly, help me here.

Reply:
Right! We should keep the epsilon in the denominator:

```python
y_true = y_true / jnp.maximum(jnp.linalg.norm(y_true, axis=axis, keepdims=True)), utils.EPSILON)
y_pred = y_pred / jnp.maximum(jnp.linalg.norm(y_pred, axis=axis, keepdims=True)), utils.EPSILON)
```

Reply:
Hy, @cgarciae  thanks for helping me out. 
I've just one query.
```
    def _l2_normalize(x, axis=None, epsilon=utils.EPSILON):
        """Normalizes along dimension `axis` using an L2 norm."""
        square_sum = jnp.sum(jnp.square(x), axis=axis, keepdims=True)
        x_inv_norm = rsqrt(jnp.maximum(square_sum, epsilon))
        return jnp.multiply(x, x_inv_norm)
```
Assume a case where ```epsilon > square_sum```
from my implementation it would be,
``` y_true = y_true / sqrt(utils.EPSILON)```
but from your approach it'll be
``` y_true = y_true / utils.EPSILON```
Please correct me if I'm wrong.
THanks.

Reply:
This is correct. Since `EPSILON` is an arbitrary constant I don't think it doesn't matters much if its has the square root or not, if you want you can use `np.sqrt(utils.EPSILON)` instead to maintain equivalence with the TF implementation.

Reply:
Thanks @cgarciae , I would like to go with ```np.sqrt(utils.EPSILON)``` . Only for the sake of maintaining TF equivalnce. Will be updating my PR with these changes very soon.